include/irr/builtin/material_compiler/glsl/rasterization/impl.glsl

```diff
@@ -3,150 +3,189 @@
-	{
-#ifdef NDF_GGX
-#ifndef ONLY_ONE_NDF
-		if (ndf==NDF_GGX) {
+#ifndef NO_TWOSIDED
+	handleTwosided(instr, s, _microfacet);
 #endif
 
-#ifdef ALL_ISOTROPIC_BXDFS
-			bxdf_eval_scalar_part = irr_glsl_ggx_height_correlated_cos_eval_DG(currBSDFParams.isotropic, currInteraction.isotropic, a2);
-#else
-			bxdf_eval_scalar_part = irr_glsl_ggx_height_correlated_aniso_cos_eval_DG(currBSDFParams, currInteraction, a, ay);
-#endif
+	const float cosFactor = irr_glsl_conditionalAbsOrMax(is_bsdf, s.NdotL, 0.0);
 
+	uvec3 regs = instr_decodeRegisters(instr);
 
-#ifndef ONLY_ONE_NDF
-		} else
-#endif
+	irr_glsl_AnisotropicMicrofacetCache microfacet;
+	bool is_valid = true;
+	bool refraction = false;
+#ifndef NO_BSDF
+	//here actually using stronger check for BSDF because it's probably worth it
+	if (op_isBSDF(op) && irr_glsl_isTransmissionPath(currInteraction.isotropic.NdotV, s.NdotL))
+	{
+		float orientedEta, rcpOrientedEta;
+		irr_glsl_getOrientedEtas(orientedEta, rcpOrientedEta, currInteraction.isotropic.NdotV, ior[0].x);
+		is_valid = irr_glsl_calcAnisotropicMicrofacetCache(microfacet, true, currInteraction.isotropic.V.dir, s.L, currInteraction.T, currInteraction.B, currInteraction.isotropic.N, s.NdotL, s.VdotL, orientedEta, rcpOrientedEta);
+		refraction = true;
+	}
+	else
 #endif
+	{
+		microfacet = _microfacet;
+	}
 
-#ifdef NDF_BECKMANN
-#ifndef ONLY_ONE_NDF
-		if (ndf==NDF_BECKMANN) {
+	bxdf_eval_t bxdf_eval = bxdf_eval_t(0.0);
+	if (is_valid && cosFactor>FLT_MIN)
+	{
+#if defined(OP_DIFFUSE) || defined(OP_DIFFTRANS)
+		if (op_isDiffuse(op))
+		{
+			vec3 reflectance = is_bsdf ? vec3(1.0) : refl;
+			float alpha2 = is_bsdf ? 0.0 : a2;
+			bxdf_eval = reflectance * irr_glsl_oren_nayar_cos_eval(s, currInteraction.isotropic, alpha2);
+		} else
 #endif
-
-#ifdef ALL_ISOTROPIC_BXDFS
-			bxdf_eval_scalar_part = irr_glsl_beckmann_smith_height_correlated_cos_eval_DG(currBSDFParams.isotropic, currInteraction.isotropic, a2);
-#else
-			bxdf_eval_scalar_part = irr_glsl_beckmann_aniso_smith_height_correlated_cos_eval_DG(currBSDFParams, currInteraction, a, a2, ay, ay2);
+#if defined(OP_CONDUCTOR) || defined(OP_DIELECTRIC)
+		if (op_hasSpecular(op))
+		{
+		BEGIN_CASES(ndf)
+#ifdef NDF_GGX
+			CASE_BEGIN(ndf, NDF_GGX) {
+				bxdf_eval_scalar_part = irr_glsl_ggx_height_correlated_aniso_cos_eval_DG(s, microfacet, currInteraction, a, a2, ay, ay2);
+			} CASE_END
 #endif
 
-#ifndef ONLY_ONE_NDF
-		} else
-#endif
+#ifdef NDF_BECKMANN
+			CASE_BEGIN(ndf, NDF_BECKMANN) {
+				bxdf_eval_scalar_part = irr_glsl_beckmann_aniso_height_correlated_cos_eval_DG(s, microfacet, currInteraction, a, a2, ay, ay2);
+			} CASE_END
 #endif
 
 #ifdef NDF_PHONG
-#ifndef ONLY_ONE_NDF
-		if (ndf==NDF_PHONG) {
+			CASE_BEGIN(ndf, NDF_PHONG) {
+				float n = irr_glsl_alpha2_to_phong_exp(a2);
+				float ny = irr_glsl_alpha2_to_phong_exp(ay2);
+				bxdf_eval_scalar_part = irr_glsl_blinn_phong_cos_eval_DG(s, microfacet, currInteraction, n, ny, a2, ay2);
+			} CASE_END
```

Comment:
if transmission occurs, how will the `cos_eval_DG` functions play with the fact that one of `VdotN` or `LdotN` will be negative? especially the smith masking function?

Furthermore at least one of `VdotH` and `LdotH` will be negative too!